sdk/samples/iot/hub/src/pnp_helper.c

```diff
@@ -0,0 +1,438 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <stdint.h>
+#include <stdio.h>
+
+#include <azure/core/az_json.h>
+#include <azure/core/az_result.h>
+#include <azure/core/az_span.h>
+#include <azure/iot/az_iot_hub_client.h>
+
+#include "pnp_helper.h"
+
+#define JSON_DOUBLE_DIGITS 2
+
+// Property values
+static char pnp_properties_buffer[64];
+
+// Device twin values
+static const az_span component_telemetry_prop_span = AZ_SPAN_LITERAL_FROM_STR("$.sub");
+static const az_span desired_temp_response_value_name = AZ_SPAN_LITERAL_FROM_STR("value");
+static const az_span desired_temp_ack_code_name = AZ_SPAN_LITERAL_FROM_STR("ac");
+static const az_span desired_temp_ack_version_name = AZ_SPAN_LITERAL_FROM_STR("av");
+static const az_span desired_temp_ack_description_name = AZ_SPAN_LITERAL_FROM_STR("ad");
+static const az_span component_specifier_name = AZ_SPAN_LITERAL_FROM_STR("__t");
+static const az_span component_specifier_value = AZ_SPAN_LITERAL_FROM_STR("c");
+static const az_span command_separator = AZ_SPAN_LITERAL_FROM_STR("*");
+static const az_span sample_iot_hub_twin_desired_version = AZ_SPAN_LITERAL_FROM_STR("$version");
+static const az_span sample_iot_hub_twin_desired = AZ_SPAN_LITERAL_FROM_STR("desired");
+
+// Visit each valid property for the component
+static az_result visit_component_properties(
+    az_span component_name,
+    az_json_reader* json_reader,
+    int32_t version,
+    char* scratch_buf,
+    int32_t scratch_buf_len,
+    pnp_helper_property_callback property_callback,
+    void* context_ptr)
+{
+  int32_t len;
+
+  while (az_succeeded(az_json_reader_next_token(json_reader)))
+  {
+    if (json_reader->token.kind == AZ_JSON_TOKEN_PROPERTY_NAME)
+    {
+      if (az_failed(az_json_token_get_string(
+              &(json_reader->token), (char*)scratch_buf, (int32_t)scratch_buf_len, (int32_t*)&len)))
+      {
+        printf("Failed to get string property value\r\n");
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+
+      if (az_failed(az_json_reader_next_token(json_reader)))
+      {
+        printf("Failed to get next token\r\n");
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+
+      if (memcmp(
```

Comment:
>memcmp [](start = 10, length = 6)

can we read more then what is valid in scratch buffer?

Reply:
updated logic to use `az_json_token_is_text_equal`